src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp

```diff
@@ -73,17 +73,22 @@ class ONNXDequantizeLinearOpLoweringToTOSA
 
     // Dequantization formula is (x - zero_point) * scale
     // Cast into the destination type first
+
+    // Cast the operands of (x - zero_point) to float32 to avoid underflows
+    Type arithType = rewriter.getF32Type();
+    Value subOpA = tosaBuilder.castToNewTensorElementType(x, arithType);
+    Value subOpB = tosaBuilder.castToNewTensorElementType(zpConst, arithType);
     Value subOp = tosa::CreateOpAndInfer<mlir::tosa::SubOp>(
-        rewriter, loc, x.getType(), x, zpConst)
+        rewriter, loc, subOpA.getType(), subOpA, subOpB)
                       .getResult();
-    Value castOp = tosa::CreateOpAndInfer<mlir::tosa::CastOp>(
-        rewriter, loc, resultType, subOp)
-                       .getResult();
     Value mulOp = tosa::CreateOpAndInfer<mlir::tosa::MulOp>(
-        rewriter, loc, resultType, castOp, scaleFactorConst, 0)
+        rewriter, loc, subOp.getType(), subOp, scaleFactorConst, 0)
                       .getResult();
+    Value castOp = tosa::CreateOpAndInfer<mlir::tosa::CastOp>(
+        rewriter, loc, resultType, mulOp)
+                       .getResult();
```

Comment:
```suggestion
    Value castOp = rewriter().createOrFold<mlir::tosa::CastOp>(
        loc, resultType, mulOp);
```
will omit the cast if its from f32 to f32

Reply:
Ah perfect! Thanks!